nullaway/src/main/java/com/uber/nullaway/generics/ConstraintSolverImpl.java

```diff
@@ -0,0 +1,303 @@
+      }
+    }
+
+    /* ---------- build final solution map ---------- */
+    Map<TypeVariable, InferredNullability> result = new HashMap<>();
+    vars.forEach(
+        (tv, st) -> {
+          // if the nullness state is UNKNOWN, set it to NONNULL arbitrarily
+          // TODO does this matter?  should we use NULLABLE instead?
+          NullnessState n =
+              (st.nullness == NullnessState.UNKNOWN) ? NullnessState.NONNULL : st.nullness;
+          result.put(
+              tv,
+              n == NullnessState.NULLABLE
+                  ? InferredNullability.NULLABLE
+                  : InferredNullability.NONNULL);
+        });
+    return result;
+  }
+
+  private void directlyConstrainTypePair(Type s, Type t) throws UnsatisfiableConstraintsException {
+    /* variable-to-variable edge */
+    if (isTypeVariable(s) && isTypeVariable(t)) {
+      TypeVariable sv = (TypeVariable) s;
+      TypeVariable tv = (TypeVariable) t;
+      getState(sv).supertypes.add(tv);
+      getState(tv).subtypes.add(sv);
+    }
+
+    /* top-level nullability rules */
+    if (isKnownNonNull(t)) {
+      requireNonNull(s);
+    }
+    if (isKnownNullable(s)) {
+      requireNullable(t);
+    }
+  }
+
+  /* ───────────────────── nullability bookkeeping ───────────────────── */
+
+  /** Force {@code tv} to {@code n}. Returns true if state changed. */
+  private boolean updateNullness(TypeVariable tv, NullnessState n)
+      throws UnsatisfiableConstraintsException {
+    VarState st = getState(tv);
+
+    if (st.nullness == n) {
+      return false;
+    }
+    if (st.nullness != NullnessState.UNKNOWN) {
+      throw new UnsatisfiableConstraintsException(
+          "Contradictory nullability for " + tv + ": " + st.nullness + " vs. " + n);
+    }
+    if (n == NullnessState.NULLABLE && !st.nullableAllowed) {
+      throw new UnsatisfiableConstraintsException(
+          tv + " cannot be @Nullable (upper bound is @NonNull)");
+    }
+    st.nullness = n;
+    return true;
+  }
+
+  private void requireNullable(Type t) throws UnsatisfiableConstraintsException {
+    if (isTypeVariable(t)) {
+      updateNullness((TypeVariable) t, NullnessState.NULLABLE);
+    } else if (isKnownNonNull(t)) {
+      throw new UnsatisfiableConstraintsException("Cannot treat @NonNull type as @Nullable: " + t);
+    }
+  }
+
+  private void requireNonNull(Type t) throws UnsatisfiableConstraintsException {
+    if (isTypeVariable(t)) {
+      updateNullness((TypeVariable) t, NullnessState.NONNULL);
+    } else if (isKnownNullable(t)) {
+      throw new UnsatisfiableConstraintsException("Cannot treat @Nullable type as @NonNull: " + t);
+    }
+  }
+
+  /* ───────────────────── helpers & stubs ───────────────────── */
+
+  private VarState getState(TypeVariable tv) {
+    return vars.computeIfAbsent(tv, v -> new VarState(upperBoundIsNullable(v)));
+  }
+
+  private static boolean isTypeVariable(Type t) {
+    // for now ignore captures
```

Comment:
Captures means `<?>` here. Right?

Reply:
Yes, I clarified the comment in 8ad8e7b6bd9195078fd80eb5d06640610cf973e4